### PR TITLE
Save notebook before running cells

### LINKED ISSUE #461: possibly unneeded peek at state? (maybe need small refactor of `saveNotebook`)
just noticed this in the code here:

https://github.com/iodide-project/iodide/blob/e1520924cdb9326aa01ed8c9f1957f0d6eafb3cf/src/task-definitions.js#L259

```
  253  
  254: tasks.saveNotebook = new UserTask({
  255    title: 'Save Notebook',
  256    keybindings: ['ctrl+s', 'meta+s'],
  257    displayKeybinding: commandKey('S'),
  258    preventDefaultKeybinding: true,
  259:   callback() { dispatcher.saveNotebook(store.getState().title) },
  260  })
  261  
```
I think we should not need that `store.getState().title` for any reason -- the reducer will have access to the store as it's processing the `saveNootbook` action.

@hamilton, do you recall any reason why this was needed? Even if it is currently needed, a small refactor of `saveNotebook` should be able to polish it out (I can take that on if it's needed)

### LINKED ISSUE #503: always autosave before executing cell
in case you hit a chunk of long running code or something that hangs and you need to bail out and reload the page, we should make it easy to recover the code you had been using previously.

i think this should be pretty easy to implement, probably just adding an additional `saveNotebook` dispatch during the cell evaluation user task

### PR DESCRIPTION
**Pull request by pradeepgangwar**
Fixes #503 
Fixes #461

### CODE REVIEW

**Comment by bcolloran** [comment]
Thanks for giving this a try @pradeepgangwar.

It does save to local storage for "ctrl+enter" and "run all", but it looks like it still needs to be implemented for "shift+enter". Here's what i tried:
- open a new dev notebook at `dev/iodide.dev.html`
- change the NB title to something i haven't used before: "newname1"
- run the first cell with "shift+enter"
- check `Object.keys(localStorage)`
- the title "newname1" does not appear in the list of saved items.

Also, this should update the *autosave* not the normal "save" entry in localStorage.

finally, if you want to try something more ambitious, you might also be able to fix #461 while you're looking at stuff related to saving :-) (but of course you don't have to try for that bigger task if you are not interested).

Thanks!

**Comment by pradeepgangwar** [comment]
@bcolloran I have updated the files and they work fine on my local machine. Evaluating cells changes autosave entry and normal 'save' entry remains untouched. Please review :smiley:

**Comment by pradeepgangwar** [comment]
@bcolloran After latest commit on this PR I also tried addressing issue #461. Please have a look into it. I found it to be working on my machine.

File: src/reducers/notebook-reducer.js [hunk 1]
```diff
@@ -89,11 +92,6 @@ const notebookReducer = (state = newNotebook(), action) => {
         }),
       }, { title: state.title === undefined ? 'new notebook' : state.title })
```
Comment (bcolloran) [inline src/reducers/notebook-reducer.js]: after your other nice refactoring, do we still need `{ title: state.title === undefined ? 'new notebook' : state.title }`? Can the title ever be undefined at this point? If not, we can remove all the checks for that, and we can even update the state schema in state-prototypes.js, the fixme around line 80:
`title: {}, // FIXME change to string ONLY with default '' or 'untitled'`

**Comment by pradeepgangwar** [comment]
@bcolloran I have added all changes suggested by you in previous comments. Please review :smiley: 
I am unable to figure out why these tests are failing? :worried:  I haven't changed anything related to those files.

File: src/task-definitions.js [hunk 2]
```diff
@@ -35,13 +35,17 @@ tasks.evaluateCell = new UserTask({
   callback() {
     dispatcher.changeMode('command')
     dispatcher.evaluateCell()
+    dispatcher.saveNotebook(true)
```
Comment (bcolloran) [inline src/task-definitions.js]: oops, we need to autosave _before_ we evaluate the cell.

**Pull request merged**